### PR TITLE
Simplify grammar and parser for observe mini-language

### PR DESCRIPTION
**Pull request by mdickinson**
This PR simplifies the grammar file and the parser for the observe mini-language, while not actually changing the set of legal expressions or their meaning.

Context: I was working on #1505, but the changes required in the grammar seemed needlessly complicated, so I started simplifying; this PR pulls out just the simplifications from that work, without any changes in semantics.

The main simplifications are: 

- the removal of the `notified_element`, `quiet_element` and `last_element` productions in the grammar, and
- the removal of the `default_notifies` stack in the parser; we only need a single flag, rather than a stack of flags

### CODE REVIEW

**Comment by rahulporuri** [comment]
These changes go over my head so i'm probably not the best person to be reviewing this PR at the moment - i'll need a lot of introduction to the grammer and parsing machinery before I can meaningfully review this PR.

File: traits/observation/_dsl_grammar.lark [hunk 1]
```diff
@@ -3,52 +3,36 @@
...
-
-// Connectors
-_dot: "."
-_colon: ":"
-
-// Grouped elements, e.g. "[a.b]", "[a,b]"
-?grouped: "[" parallel "]"
-
 // Simple name as trait name, e.g. "a"
-trait: NAME -> trait
```
Comment (mdickinson) [inline traits/observation/_dsl_grammar.lark:16]: The "-> trait" part here is an alias, but it's aliasing "trait" to itself. I removed this and other self-aliasing cases, to make the grammar easier to parse mentally without needing to know about this particular lark feature.

File: traits/observation/_dsl_grammar.lark [hunk 2]
```diff
@@ -3,52 +3,37 @@
...
 // Rule for joining elements in series, e.g. "a.b.c"
-?series: (notified_element | quiet_element)* last_element
+?series: (series (notify | quiet))? element
 
 // Rule for joining elements in parallel, e.g. "a,b,c"
-?parallel: series ("," series)*
+?parallel: (parallel ",")? series
 
-// Final expression to be used in start
-?expr: parallel
```
Comment (mdickinson) [inline traits/observation/_dsl_grammar.lark:51]: We have `start: expr` and `expr: parallel`, and that's the only place that `expr` is used, so I short-circuited that to `start: parallel`.

**Review by rahulporuri — APPROVED** [review]
I'd still love a one-on-one chat to discuss what was changed in this PR and why but the changes LGTM for now because I have faith in the testsuite. As I understand it, there are other changes that rely on this PR so I don't want to block future/planned work.

**Comment by mdickinson** [comment]
> I'd still love a one-on-one chat to discuss what was changed in this PR and why

I'd be happy to do that any time! Just ping me.

**Comment by mdickinson** [comment]
From external discussion: it would be good to have a few more tests, and particularly some tests that properly exercise the notify stack that's in the current main branch, so we can be sure that those cases are still dealt with correctly in the new code.

**Comment by mdickinson** [comment]
Added a couple of tests, and fixed merge conflicts (twice!). Will merge when CI completes.

**Pull request merged**